### PR TITLE
replace_symbolt refactoring and stricter type checking

### PR DESCRIPTION
**Pull request by tautschnig**
The first three commits are in separate pull requests already (#2720, #2721, #2722). The last three commits are actually new. Please provide feedback whether this design is acceptable/better (@kroening in particular).

### CODE REVIEW

**Comment by tautschnig** [comment]
Further cleanup and refactoring based on feedback provided by @kroening - I think this is much cleaner and simpler now. Thanks a lot!

**Review by kroening — APPROVED** [review]

File: src/goto-instrument/concurrency.cpp [hunk 1]
```diff
@@ -89,19 +89,17 @@ void concurrency_instrumentationt::instrument(exprt &expr)
...
+      const symbol_exprt &s = to_symbol_expr(*s_it);
 
-      shared_varst::const_iterator
-        v_it=shared_vars.find(identifier);
+      shared_varst::const_iterator v_it = shared_vars.find(s.get_identifier());
 
       if(v_it!=shared_vars.end())
       {
         index_exprt new_expr;
         // new_expr.array()=symbol_expr();
```
Comment (peterschrammel) [inline src/goto-instrument/concurrency.cpp:99]: remove commented code

File: src/util/replace_symbol.cpp [hunk 2]
```diff
@@ -252,3 +248,85 @@ void unchecked_replace_symbolt::insert(
...
+    if(!unchecked_replace_symbolt::replace(dest.type()))
+      result=false;
+  }
+
+  // now do expression itself
+
+  if(!have_to_replace(dest))
+    return result;
+
+  if(dest.id()==ID_index)
```
Comment (peterschrammel) [inline src/util/replace_symbol.cpp]: clang-format

**Review by peterschrammel — APPROVED** [review]

File: src/goto-instrument/concurrency.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/concurrency.cpp:99]: Will do in a separate PR

**Pull request merged**